### ISSUE TITLE
Extend `Style/SelfAssignment` beyond variables

### ISSUE THREAD

**Issue by vlad-pisanov** [rubocop/rubocop#12410]
Is there a reason why `Style/SelfAssignment` only works with variables, as opposed to general expressions? I.e. could we lint for something like:

```ruby
# bad
hash['foo'] = hash['foo'] + 42
user.name = user.name + " Jr."

# good
hash['foo'] += 42
user.name += " Jr."
```

**Comment by ydakuka** [comment]
And

```ruby
# bad

x = x % 2
x = x ^ 2
x = x << 2
x = x >> 2
```

```ruby
# good

x %= 2
x ^= 2
x <<= 2
x >>= 2
```

**Comment by koic** [comment]
The main issue was resolved by #12415, and the derivative will be resolved by #12421.
